Pda messaging attempt

PolterTzi opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Admittedly likely messy attempt at adding PDA messaging.
Surely needs cleanup and adjustments.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
PDA messaging would offer a great benefit to inter-player communication and convenience.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Adds a Messages app cartridge to the PDA and a MessagesServer component. 
When the server is powered, it attempts to download the names of users from the cartridges.
Whenever a name changes or a new name is added, it attempts to broadcast that name to all cartridges on the same map,.
The app can be used to send messages to other players. The messages are sent in a struct containing the sender and recipient ids (based on the entity id of their idcards) to the first server the cartridge can find. The server then attempts to send the message on to the recipient.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

https://github.com/space-wizards/space-station-14/assets/62557990/75eda7c9-91e1-43d9-9419-034e37ce450d



- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

:cl:
- add: Added basic PDA messaging functionality.

Kadeo64 commented:
huge W

Kadeo64 commented:
anyway before maintainers get here to mention it you should follow the [design document](https://docs.spacestation14.com/en/proposals/julian-vasilis-pda-messaging.html) that's been accepted for PDA messaging.

PolterTzi commented:
> anyway before maintainers get here to mention it you should follow the [design document](https://docs.spacestation14.com/en/proposals/julian-vasilis-pda-messaging.html) that's been accepted for PDA messaging.

The document seems generally wide enough that I think I'd be reasonably within the margins of minimal requirements.

metalgearsloth commented on Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:
```diff
@@ -319,6 +356,23 @@ protected override void OnItemInserted(EntityUid uid, CartridgeLoaderComponent l
 
     protected override void OnItemRemoved(EntityUid uid, CartridgeLoaderComponent loader, EntRemovedFromContainerMessage args)
     {
+        if (args.Container.ID == PdaComponent.PdaIdSlotId)
+        {
+            //_sawmill.Debug("id removal detected");
+            if (TryComp(uid, out ContainerManagerComponent? cont))
+            {
+                //_sawmill.Debug("trying program");
+                if (TryGetProgram<MessagesCartridgeComponent>(uid, out _, out var messagesCartridgeComponent, false, loader, cont))
+                {
+                    //_sawmill.Debug("got program, updating data");
+                    messagesCartridgeComponent.ConnectedId = null;
+                    messagesCartridgeComponent.UserUid = null;
+                    messagesCartridgeComponent.UserName = null;
+                    //_sawmill.Debug("data updated");
+                }
+            }
+        }
+
         if (args.Container.ID != InstalledContainerId && args.Container.ID != loader.CartridgeSlot.ID)
```
Same here don't just ignore the event and hardcode in a specific cartridge.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeComponent.cs:
```diff
@@ -0,0 +1,66 @@
+    /// The name and job of the user
+    /// </summary>
+    [DataField]
+    public string? UserName = null;
+
+     /// <summary>
+    /// The uid of the crew the user is chatting with
+    /// </summary>
+    [DataField]
+    public string? ChatUid = null;
+
+    /// <summary>
+    /// ID card connected to the Cartridge
+    /// </summary>
+    [DataField]
+    public EntityUid? ConnectedId = null;
+
+    /// <summary>
+    /// Dictionary translating uids to readable names
+    /// </summary>
+    [DataField]
+    public Dictionary<string,string> NameDict= new();
+
+    /// <summary>
+    /// Whether the cartridge has lost connection and should be looking for a new server
+    /// </summary>
+    [DataField]
+    public bool DeadConnection=true;
+
+    /// <summary>
+    /// Key used to denote which faction the cartridge belongs to.
+    /// </summary>
+    [DataField]
+    public string EncryptionKey = "Nanotrasen";
+
+    public TimeSpan UpdateDelay = TimeSpan.FromSeconds(10);
+    public TimeSpan NextUpdate = TimeSpan.Zero;
+
+
+}
```
Don't use magic strings everywhere because they break easily. Use prototypes or proper data structures.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+    [Dependency] private readonly CartridgeLoaderSystem _cartridgeLoaderSystem = default!;
+    [Dependency] private readonly MessagesServerSystem _messagesServerSystem = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly IEntitySystemManager _entitySystem = default!;
```
Just cache the system directly.

PolterTzi replied:
I honestly lack the knowledge to understand what exactly you mean.

metalgearsloth replied:
You never need IEntitySystemManager on a system.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+                if (idCardComponent.FullName != "")
+                {
+                    component.UserName = idCardComponent.FullName+"("+idCardComponent.JobTitle+")";
+                }
+                else
+                {
+                    component.UserName = "Unknown "+"("+idCardComponent.JobTitle+")";
+                }
+            }
+            if (component.UserUid != null && component.UserName != null)
+                component.NameDict[component.UserUid] = component.UserName;
+        }
+
+        //_sawmill.Debug("Updating cart");
+        if ((component.MessagesQueue.Count > 0) || component.DeadConnection)
+        {
+            if (!(GetActiveServer(component, mapId) is var (serverUid, messageServer)))
+            {
+                component.DeadConnection=true;
+                return;
+            }
+            else
+            {
+                if (component.DeadConnection)
+                    PullFromServer(uid, component, messageServer);
+                component.DeadConnection=false;
+            }
+
+
+            var tempMessageQueue = new List<MessagesMessageData>(component.MessagesQueue);
+            foreach (var message in tempMessageQueue)
+            {
+                _messagesServerSystem.PdaToServerMessage(serverUid, messageServer,message);
+                component.MessagesQueue.Remove(message);
+                component.Messages.Add(message);
+                UpdateUiState(uid, Transform(uid).ParentUid, component);
+            }
+        }
+    }
+
```
1. Don't use string concat everywhere as mentioned above.
2. Run a formatter over this.
3. Don't just hardcode in getting the username every tick.
4. This isn't localised.
5. If you need a cartridge to poll add a dedicated component instead of having hundreds of cartridges just polling for servers.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+    public (EntityUid, MessagesServerComponent)? GetActiveServer(MessagesCartridgeComponent component,MapId mapId)
+    {
+       // _sawmill.Debug("Looking for server");
+        var servers = EntityQueryEnumerator<MessagesServerComponent, ApcPowerReceiverComponent, TransformComponent>();
```
You should be using AllEntityQuery not this.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+       // _sawmill.Debug("No server");
+        return null;
+    }
+
+    private string GetName(MessagesCartridgeComponent component, string key)
+    {
+        if (!(component.NameDict.ContainsKey(key)))
+        {
+            return "Unknown";
+        }
+        else
+        {
+            if (component.NameDict[key][0] == '(')
+            {
+                return "Unknown "+component.NameDict[key];
+            }
+            else
+            {
+                return component.NameDict[key];
+            }
+        }
+    }
+
```
As above. To be honest just use identity to get the name at runtime instead of having to fix this properly.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+
+            userList.Sort(delegate((string,string) a, (string,string) b)
+            {
+                    return String.Compare(a.Item2,b.Item2);
+            });
+
+            state = new MessagesUiState(MessagesUiStateMode.UserList, userList, null);
+        }
+        else
+        {
+            List<MessagesMessageData> messageList = new();
+
+            foreach (var message in component.Messages)
+            {
+                if ((message.SenderId == component.ChatUid && message.ReceiverId == component.UserUid) || (message.ReceiverId == component.ChatUid && message.SenderId == component.UserUid))
+                {
+                    messageList.Add(message);
+                }
+            }
+
+            messageList.Sort(delegate(MessagesMessageData a, MessagesMessageData b)
+            {
+                return TimeSpan.Compare(a.Time,b.Time);
+            });
+
+            List<(string, string)> formattedMessageList = new();
+
+            foreach (var message in messageList)
+            {
+                string name = GetName(component, message.SenderId);
+                var stationTime =message.Time.Subtract(_gameTicker.RoundStartTimeSpan);
+                string content = stationTime.ToString("\\[hh\\:mm\\:ss\\]")+" : "+message.Content;
+                formattedMessageList.Add((name , content));
+            }
+
+            state = new MessagesUiState(MessagesUiStateMode.Chat, formattedMessageList, GetName(component, component.ChatUid));
+        }
+        _cartridgeLoaderSystem?.UpdateCartridgeUiState(loaderUid, state);
+    }
+
```
As mentioned above configure your IDE to provide warnings because there's a bunch of dummy code here such as the null operator that is unnecessary.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+        _cartridgeLoaderSystem?.UpdateCartridgeUiState(loaderUid, state);
+    }
+
+    public void ServerToPdaMessage(EntityUid uid, MessagesCartridgeComponent component, MessagesMessageData message, EntityUid pdaUid, MessagesServerComponent server)
+    {
+        //_sawmill.Debug("Message received - cart");
+        component.Messages.Add(message);
+
+        string name = GetName(component, message.SenderId);
+
+        var wrappedMessage = Loc.GetString("chat-radio-message-wrap",
+            ("color", Color.White),
+            ("fontType", "Default"),
+            ("fontSize", 12),
+            ("verb", "sends"),
+            ("channel", "[PDA]"),
+            ("name", name),
+            ("message", message.Content));
+
+        var chat = new ChatMessage(
+            ChatChannel.Radio,
+            message.Content,
+            wrappedMessage,
+            NetEntity.Invalid,
+            null);
+
+        var chatMsg = new MsgChatMessage { Message = chat };
+
+        if (TryComp(Transform(pdaUid).ParentUid, out ActorComponent? actor))
+            _netMan.ServerSendMessage(chatMsg, actor.PlayerSession.Channel);
+        if (TryComp(pdaUid, out RingerComponent? ringer))
+            _ringer.RingerPlayRingtone(pdaUid, ringer);
+
+
+        if (TryComp(pdaUid, out CartridgeLoaderComponent? loaderUid))
+            UpdateUiState(uid, pdaUid, component);
+    }
+
```
You should not be hardcoding in a check for the PDA's parent because this breaks in many situations.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,310 @@
+            UpdateUiState(uid, pdaUid, component);
+    }
+
+
+    public void PullFromServer(EntityUid uid, MessagesCartridgeComponent component, MessagesServerComponent server)
+    {
+        //_sawmill.Debug("Syncing with server");
+
+        if (component.ConnectedId == null)
+            return;
+
+        if (!(TryComp(component.ConnectedId, out IdCardComponent? idCardComponent)))
+            return;
+
+        component.UserName = idCardComponent.FullName+"("+idCardComponent.JobTitle+")";
+
+        if (server.NameDict != component.NameDict)
+        {
+            if ((component.UserUid is string userUid) && (component.UserName is string userName))
+                server.NameDict[userUid]=userName;
+            var keylist = server.NameDict.Keys;
+            foreach (var key in keylist)
+            {
+                component.NameDict[key]=server.NameDict[key];
+            }
+        }
+
+        //_sawmill.Debug("Pulling messages from server");
+
+        foreach (var message in server.Messages)
+        {
+            if ((message.ReceiverId == component.UserUid || message.SenderId == component.UserUid) && !(component.Messages.Contains(message)))
+                component.Messages.Add(message);
+        }
+
+
```
As above on string concat + name handling, also please document your code.

metalgearsloth commented on Content.Server/Radio/Components/MessagesServerComponent.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Content.Server.Radio.Components;
+
+/// <summary>
+/// Entities with <see cref="MessagesServerComponent"/> are needed to transmit messages using PDAs.
+/// They also need to be powered by <see cref="ApcPowerReceiverComponent"/>
+/// in order for them to work on the same map as server.
+/// </summary>
+[RegisterComponent]
+public sealed partial class MessagesServerComponent : Component
+{
+    /// <summary>
+    /// The list of messages cached by the server.
+    /// </summary>
+    [DataField]
+    public List<MessagesMessageData> Messages = new();
+
+    /// <summary>
+    /// The list of messages the server is trying to send.
+    /// </summary>
+    [DataField]
+    public List<MessagesMessageData> MessagesQueue = new();
+
+    /// <summary>
+    /// Dictionary translating uids to readable names
+    /// </summary>
+    [DataField]
+    public Dictionary<string,string> NameDict= new();
+
+    /// <summary>
+    /// Key used to denote which faction the server belongs to.
+    /// </summary>
+    [DataField]
+    public string EncryptionKey = "Nanotrasen";
+
+    public TimeSpan UpdateDelay = TimeSpan.FromSeconds(3);
+    public TimeSpan NextUpdate = TimeSpan.Zero;
+    public TimeSpan SyncDelay = TimeSpan.FromSeconds(180);
+    public TimeSpan NextSync = TimeSpan.Zero;
+}
```
You're storing several copies of each message + name + faction + update in a copy-paste format which is just going to make this a pain to maintain down the line.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        if ((component.MessagesQueue.Count > 0))
+        {
+            var tempMessagesQueue = new List<MessagesMessageData>(component.MessagesQueue);
+            foreach (var message in tempMessagesQueue)
+            {
+                bool sent=TryToSend(message, mapId, component);
+                if (sent)
+                {
+                    component.MessagesQueue.Remove(message);
+                    component.Messages.Add(message);
+                }
+            }
+        }
+    }
+
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent>();
+        while(query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if ((cartComponent.UserUid == null) || (cartComponent.UserName == null))
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
```
Having polling in both PDAs and server is unnecessary, you should just be doing it on the server to keep the code significantly cleaner than this and the copied checks everywhere.

Additionally this code must be able to handle generic messengers (for example PAIs) and not just be hardcoded to PDAs.

metalgearsloth replied:
See my comment below on:
- The caching handling looks really dodgy.
- You should be storing what components are valid up-front and not re-iterating it every time for TrySend.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    public bool TryToSend(MessagesMessageData message, MapId mapId, MessagesServerComponent server)
+    {
+        bool sent = false;
+
+        var query = EntityQueryEnumerator<CartridgeLoaderComponent, ContainerManagerComponent>();
+
+        while (query.MoveNext(out var uid, out var comp, out var cont))
+        {
+            if (!_cartridgeLoaderSystem.TryGetProgram<MessagesCartridgeComponent>(uid, out var progUid, out var messagesCartridgeComponent, false, comp, cont))
+                continue;
+            if (progUid is EntityUid realProgUid)
+            {
+                if (messagesCartridgeComponent.UserUid == message.ReceiverId)
+                    _messagesCartridgeSystem.ServerToPdaMessage(realProgUid, messagesCartridgeComponent, message, uid, server);
+                sent = true;
+            }
+        }
+
+        return sent;
+    }
+
+    public void PdaToServerMessage(EntityUid uid, MessagesServerComponent component, MessagesMessageData message)
+    {
+        //component.Messages.Add(message);
+        component.MessagesQueue.Add(message);
+        Update(uid, component);
+    }
+
```
Just iterate the component directly, there's no reason to be checking every single loader.

metalgearsloth reviewed (changes requested):
Please cleanup all of your dummy logging and check your IDE for warnings.

PolterTzi commented on Content.Client/CartridgeLoader/Cartridges/MessagesUiFragment.xaml.cs:
```diff
@@ -0,0 +1,121 @@
+            }
+        }
+    }
+
```
HeaderBox and Input are supposed to be reused, but the contents of MessageContainer are supposed to be discarded.
I think that's what would be happening, but I could be wrong here.

PolterTzi replied:
And I'm not sure how to write this properly to avoid the contains call.

PolterTzi commented on Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:
```diff
@@ -310,6 +317,36 @@ public void UnregisterBackgroundProgram(EntityUid loaderUid, EntityUid cartridge
+            }
+        }
+
         if (args.Container.ID != InstalledContainerId && args.Container.ID != loader.CartridgeSlot.ID)
```
My issue here is I want the cartridge to react to the id being inserted into the PDA.
From what I understand the local event raised in that case reaches the cartridgeloader, but since the cartridge itself seems to be a separate entity, I don't think it would receive the event.
One way I could see doing this without hardcoding the cartridge would be to have a generic "ID was inserted" function in the cartridges that I would then call for each cartridge.

Brandon-Huu replied:
five levels of indention is insane

metalgearsloth replied:
> My issue here is I want the cartridge to react to the id being inserted into the PDA. From what I understand the local event raised in that case reaches the cartridgeloader, but since the cartridge itself seems to be a separate entity, I don't think it would receive the event. One way I could see doing this without hardcoding the cartridge would be to have a generic "ID was inserted" function in the cartridges that I would then call for each cartridge.

Add a new event to cartidges that get loaded or something, this indentation makes the entire thing hard to manage and prevent bugs.

Brandon-Huu commented on Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:
```diff
@@ -329,6 +335,28 @@ public void SendNotification(EntityUid loaderUid, string header, string message,
 
     protected override void OnItemInserted(EntityUid uid, CartridgeLoaderComponent loader, EntInsertedIntoContainerMessage args)
     {
+        if (args.Container.ID == PdaComponent.PdaIdSlotId) //Very ugly hardcoded thing to update messager app when ID card is inserted. <Todo> find some better way to do this.
+        {
+            if ((TryComp(uid, out ContainerManagerComponent? cont)))
+            {
+                if (TryGetProgram<MessagesCartridgeComponent>(uid, out var programUid, out var messagesCartridgeComponent, false, loader, cont))
+                {
+                    if (TryComp(uid, out PdaComponent? _) && programUid is EntityUid realProgramUid)
+                    {
+                        if (TryComp(args.Entity, out IdCardComponent? _))
+                        {
+                            messagesCartridgeComponent.ConnectedId = args.Entity;
+                            messagesCartridgeComponent.UserUid = args.Entity.Id;
+                            if (_messagesCartridgeSystem.GetActiveServer(messagesCartridgeComponent, Transform(uid).MapID) is var (_, server))
+                            {
+                                _messagesCartridgeSystem.PullFromServer(realProgramUid, messagesCartridgeComponent, server);
+                            }
+                        }
+                    }
+                }
+            }
+        }
```
Use guard statements

Brandon-Huu commented:
Why close?

Jezithyr commented:
Reopened since the repoban was in error. (The ban was made due to a misunderstood reference to the stonewall riots in an april fools PR)

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,166 @@
+            }
+        }
+    }
+
+    //Sync function that updates the name dictionaries of all carts to match the server.
+    //Called periodically with update.
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = _entManager.AllEntityQueryEnumerator<MessagesCartridgeComponent>();
+        while (query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if (cartComponent.UserUid == null || cartComponent.UserName == null)
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict.ContainsKey(entry.Key) || cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
+    //function that tries to send a message to any matching cartridges on its map
```
I don't really get why we even need 2 steps? I get one only applies if it has power but just seems like unnecessary iteration when we could just wrap it into the 1 update loop and save a lot of performance + headaches.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    //function that tries to send a message to any matching cartridges on its map
+    public bool TryToSend(MessagesMessageData message, MapId mapId)
+    {
+        bool sent = false;
+
+        var query = EntityQueryEnumerator<MessagesCartridgeComponent, CartridgeComponent>();
+
+        while (query.MoveNext(out var uid, out var messagesCartridgeComponent, out var cartridge))
+        {
+            if (Transform(uid).MapID != mapId)
+                continue;
+            if (cartridge.LoaderUid != null) //<TODO> this should probably be more generalisable
+            {
+                if (messagesCartridgeComponent.UserUid == message.ReceiverId)
+                    _messagesCartridgeSystem.ServerToPdaMessage(uid, messagesCartridgeComponent, message, cartridge.LoaderUid.Value);
+                sent = true;
+            }
+        }
+
+        return sent;
+    }
```
Okay having read this after the above you should just be caching what message cartridges are actually relevant for a server up-front, then re-using that data throughout the rest of the update. As it is if we have 10 messages to send you iterate every single message cartridge 10 times instead of the once per relevant cartridge.

metalgearsloth replied:
I don't think the server should really care if anyone received it? Would be better if it always just marked it as sent on an update and if no devices are connected then too bad.

metalgearsloth commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiMessageEvent.cs:
```diff
@@ -0,0 +1,27 @@
+        StringInput = stringInput;
+    }
+}
+
+[Serializable, NetSerializable]
+public enum MessagesUiAction
+{
```
```suggestion

[Serializable, NetSerializable]
public enum MessagesUiAction : byte
{
```

metalgearsloth commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiState.cs:
```diff
@@ -0,0 +1,34 @@
+﻿using Robust.Shared.Serialization;
+
+namespace Content.Shared.CartridgeLoader.Cartridges;
+
+[Serializable, NetSerializable]
+public sealed class MessagesUiState(MessagesUiStateMode mode, List<(string, int?)> contents, string? name = null) : BoundUserInterfaceState
+{
+    public List<(string, int?)>? Contents = contents;
+    public MessagesUiStateMode Mode = mode;
+    public string? Name = name;
+}
+
```
Please document.

metalgearsloth commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiState.cs:
```diff
@@ -0,0 +1,34 @@
+    public string? Name = name;
+}
+
+[Serializable, NetSerializable]
+public enum MessagesUiStateMode
+{
+    UserList,
+    Chat
+}
+
+[Serializable, NetSerializable]
+public partial struct MessagesMessageData
+{
+    public int SenderId;
+    public int ReceiverId;
+    public string Content;
+    public TimeSpan Time;
+}
+
+[Serializable, NetSerializable]
+public enum MessagesKeys
+{
+    Nanotrasen,
+    Syndicate
```
```suggestion
[Serializable, NetSerializable]
public enum MessagesUiStateMode : byte
{
    UserList,
    Chat
}

[Serializable, NetSerializable]
public partial struct MessagesMessageData : byte
{
    public int SenderId;
    public int ReceiverId;
    public string Content;
    public TimeSpan Time;
}

[Serializable, NetSerializable]
public enum MessagesKeys : byte
{
    Nanotrasen,
    Syndicate
```

Fildrance commented:
Is this halted due to chatfactor? Maybe some assistance needed?

PolterTzi commented:
> Is this halted due to chatfactor? Maybe some assistance needed? @PolterTzi

Mostly halted due to me failing to actually work on it. Chatfactor shouldn't be an issue, I think? The only part of the PR interacting with chat should be the notification (which shouldn't cause issues), unless that changes.
I'm frankly not good with organisational stuff and time management, so honestly I haven't gotten around to this since sloth commented.

PolterTzi commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeComponent.cs:
```diff
@@ -0,0 +1,58 @@
+    /// </summary>
+    [DataField]
+    public string? UserName = null;
+
+    /// <summary>
+    /// The uid of the crew the user is chatting with
+    /// </summary>
+    [DataField]
+    public int? ChatUid = null;
+
```
Yeah, bad naming (and documentation) on my part.
It's supposed to be the id of the given user (or rather ID card) in the messaging system, based on the ID card's UID.

PolterTzi commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeComponent.cs:
```diff
@@ -0,0 +1,58 @@
+    [DataField]
+    public EntityUid? ConnectedId = null;
+
+    /// <summary>
+    /// Dictionary translating uids to readable names
+    /// </summary>
+    [DataField]
+    public Dictionary<int, string> NameDict = [];
+
```
I really went hard on storing things in the cartridge itself, yeah, I'll try to remedy that.

PolterTzi replied:
I'll admit, it'll be much cleaner without that.

PolterTzi commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeComponent.cs:
```diff
@@ -0,0 +1,58 @@
+    /// </summary>
+    [DataField]
+    public MessagesKeys EncryptionKey = MessagesKeys.Nanotrasen;
+
```
Could you please point me where I could find those prototypes?

juliangiebel reviewed (changes requested):
This needs to use device networking and follow the design doc for pda messaging more closely.
I'm also not sure on the encryption key stuff.
I'll do a proper review when this gets changed to use device networking and instead of the closest server use the `SingletonDeviceNetServerComponent.cs` Content.Server/DeviceNetwork/Components/SingletonDeviceNetServerComponent.cs

PolterTzi commented:
> This needs to use device networking and follow the design doc for pda messaging more closely. I'm also not sure on the encryption key stuff. I'll do a proper review when this gets changed to use device networking and instead of the closest server use the `SingletonDeviceNetServerComponent.cs` Content.Server/DeviceNetwork/Components/SingletonDeviceNetServerComponent.cs

I've not had experience with device networking yet, so I don't know how much effort it would take to make the change. May I ask for a more precise recommendation on how to use it? I do hope it won't be too complicated to stop this from being ready in time.

juliangiebel commented:
This is the documentation for device networking: https://docs.spacestation14.com/en/space-station-14/core-tech/device-network.html
You can look at crew monitor servers for how they handle the singleton server component.

PolterTzi commented:
Thanks.
Also, could you tell me what isn't following the doc here?
I thought I followed enough to pass for, well, bare minimum viable product, but if there's some particular mistake there, I'd like to know.

juliangiebel commented:
Not using device networking was actually the biggest point that was missing. I can take a better look when that's done but it should be fine by then.

PolterTzi commented:
Ok, so, to verify.
I want to have the messages sent and received on the current map.
I think the NetId would be Wireless?
Both the cart and server should have DeviceNetworkComponent and StationLimitedNetwork. Would that be correct?

Haven't looked at how to handle the singleton yet, will in a bit.

Also, the whole "encryption key" issue could be quite cleanly be covered by frequencies, so that's one less oddity.

PolterTzi commented on Content.Client/CartridgeLoader/Cartridges/MessagesUiFragment.xaml:
Localise the ui back, make the loc string fit the rest.

PolterTzi commented on Resources/Prototypes/Device/devicenet_frequencies.yml:
Fix sprites having panel open

PolterTzi commented on Content.Client/CartridgeLoader/Cartridges/MessagesUiFragment.xaml:
Figure out why it throws an error

PolterTzi commented on Content.Client/CartridgeLoader/Cartridges/MessagesUiFragment.xaml.cs:
the back button needs to hide

juliangiebel commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,271 @@
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly MessagesServerSystem _messagesServerSystem = default!;
+    [Dependency] private readonly DeviceNetworkSystem _deviceNetworkSystem = default!;
```
You need to subscribe to the `CartridgeDeviceNetPacketEvent` instead of the `DeviceNetworkPacketEvent` as well as subscribe to the `CartridgeAddedEvent` and call  `_cartridgeLoaderSyste.RegisterBackgroundProgram`. 

This will cause the cartridge to use the PDAs device network connection instead of its own, and it will ensure that the cartridge is only connected to the network when it's installed or inserted inside a PDA.
You will need to use the `PDA` frequency for messaging which means you will have to check if it's a syndicate cartridge or not. I'd just use an int called faction or something. Calling it encryption key is confusing.

If the cartridge has its own device network connection it'll still be connected when not inserted into a PDA. You can listen to the `CartridgeActivatedEvent` and `CartridgeDeactivatedEvent` and connect and disconnect from the network there if you really want to have the cartridge use its own connection.

PolterTzi replied:
I think I'll go cartridges having their own connections.
For one thing, I personally like it better to use a unique frequency, so unless it's for some reason needed otherwise, I'll stick with that.
For another, it's gonna make it a bit easier to potentially expand the network with non-pda messagers in the future.
Either way, thank you for the instructional feedback.

juliangiebel commented on Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:
```diff
@@ -9,6 +9,9 @@
 using Robust.Shared.Player;
+using Content.Shared.PDA;
+using Content.Server.CartridgeLoader.Cartridges;
+using Content.Shared.Access.Components;
```
Remove the 3 usings above. They're not needed

juliangiebel commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiState.cs:
```diff
@@ -0,0 +1,39 @@
+/// Data of a single message in the system, containing the ids of the sender and recipient, the text content and the time it was sent.
+///</summary>
+[Serializable, NetSerializable]
+public partial struct MessagesMessageData
```
This doesn't need to be partial

juliangiebel commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,271 @@
+                Content = messageEvent.StringInput,
+                Time = _gameTiming.CurTime.Subtract(_gameTicker.RoundStartTimeSpan)
+            };
+            if (!TryComp(uid, out DeviceNetworkComponent? device))
```
Don't `TryComp` components of systems you're calling. That is taken care of in the methods of the system.

PolterTzi replied:
So it's gonna work correctly if I don't mention the device component in the method call, right?

juliangiebel replied:
Yes

juliangiebel commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,271 @@
+            {
+                ["Message"] = messageData
+            };
+            _deviceNetworkSystem.QueuePacket(uid, null, packet, device: device);
```
Don't broadcast the packet. Use `_singletonServerSystem.TryGetActiveServerAddress` to get the address of the active server.

juliangiebel reviewed (changes requested):
This is a partial review. I'll be doing the rest later.

PolterTzi commented:
With the combination of the approaching freeze and not being in the _best_ state to work on this, I unfortunately won't finish this today. 
Thank you to all the people who provided feedback, apologies to those I've disappointed it annoyed in the process and here's hoping I'll be better at this come June.

AJCM-git reviewed (commented):
gonna give you more time to finish this during the freeze since it's pretty close to being done and there is not that much work left

PolterTzi commented:
Thank you for the opportunity, I've believe I've addressed most review points by now.
Still need to check it in actual gameplay, but will do so soon.

metalgearsloth commented on Content.Client/CartridgeLoader/Cartridges/MessagesUi.cs:
```diff
@@ -0,0 +1,39 @@
+
+public sealed partial class MessagesUi : UIFragment
+{
+    private MessagesUiFragment _fragment;
+
+    public override Control GetUIFragmentRoot()
+    {
+        return _fragment!;
+    }
+
+    public override void Setup(BoundUserInterface userInterface, EntityUid? fragmentOwner)
+    {
+        _fragment = new MessagesUiFragment();
+        _fragment.OnMessageSent += note => SendMessagesMessage(MessagesUiAction.Send, note, null, userInterface);
+        _fragment.OnButtonPressed += userUid => SendMessagesMessage(MessagesUiAction.ChangeChat, null, userUid, userInterface);
+    }
+
+    public override void UpdateState(BoundUserInterfaceState state)
+    {
+        if (state is not MessagesUiState messagesState)
+            return;
+
+        _fragment?.UpdateState(messagesState.Mode, messagesState.Contents, messagesState.Name);
+    }
```
If fragment is not null anymore you don't need the null error handling code anymore so please remove.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeComponent.cs:
```diff
@@ -0,0 +1,20 @@
+
+[RegisterComponent]
+public sealed partial class MessagesCartridgeComponent : Component
+{
+    /// <summary>
+    /// The component of the last contacted server
+    /// </summary>
+    [DataField]
+    public MessagesServerComponent? LastServer = null;
+
```
You cannot store components in datafields, this will break serialization and it's also not clear why this is needed.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+    /// </summary>
+    private void OnUiReady(EntityUid uid, MessagesCartridgeComponent component, CartridgeUiReadyEvent args)
+    {
+        var stationId = _stationSystem.GetOwningStation(uid);
+        if (stationId.HasValue)
+            _singletonServerSystem.TryGetActiveServerAddress<MessagesServerComponent>(stationId.Value, out _);
+        UpdateUiState(uid, args.Loader, component);
+    }
```
This hasvalue is unnecessary just pass the nullable entityuid in.

PolterTzi replied:
TryGetActiveServerAddress doesn't seem to take a nullable argument, so I'm not sure what you mean.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+    }
+
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+        var serverQuery = EntityQueryEnumerator<MessagesServerComponent>();
+        while (serverQuery.MoveNext(out var uid, out var server))
+        {
+            if (!_singletonServerSystem.IsActiveServer(uid))
+                continue;
+            if (server.NextUpdate <= _gameTiming.CurTime)
+            {
+                server.NextUpdate += server.UpdateDelay;
+
+                Update(uid, server);
+            }
+        }
+    }
+
```
Why does a messaging server need to poll every second? Surely it can just subscribe to messages that come in instead?

PolterTzi replied:
It's supposed to update every 10 seconds to broadcast a message asking for names of users on the network so it can send the list of other users to the clients. I thought that would be better than checking names on all carts every time someone opens the program. But, now that I think about it, that still seems like something to be gotten when needed. I'll update to that.
It also used the update function to sync with other servers, but I think I can just move that to the SendMessage function.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+            component.Messages = new List<MessagesMessageData>(component.Messages.Union(serverComponent.Messages));
+            serverComponent.Messages = new List<MessagesMessageData>(component.Messages);
+        }
+
+        var packet = new NetworkPayload()
+        {
+            ["NameQuery"] = true,
+            ["ServerComponent"] = component
+        };
+        _deviceNetworkSystem.QueuePacket(uid, null, packet);
```
Please comment your code because this looks bad and I have no idea what it's trying to do so can't advise on how to fix it.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+    {
+        if (!this.IsPowered(uid, EntityManager))
+            return;
+
+        var serverQuery = EntityManager.AllEntityQueryEnumerator<MessagesServerComponent>();
+
+        while (serverQuery.MoveNext(out var serverUid, out var serverComponent))
+        {
+            component.Messages = new List<MessagesMessageData>(component.Messages.Union(serverComponent.Messages));
+            serverComponent.Messages = new List<MessagesMessageData>(component.Messages);
+        }
+
```
Instead of just doing copies to every single server can't you just store the messages on some abstract entity instead? It will be hundreds of times faster than what this code is doing.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+    {
+        if (!_singletonServerSystem.IsActiveServer(uid))
+            return;
+        if (args.Data.TryGetValue<string>("NewName", out var name) && args.Data.TryGetValue<int>("UserId", out var userId))
+            component.NameDict[userId] = name;
+        if (args.Data.TryGetValue<MessagesMessageData>("Message", out var message))
+            SendMessage(uid, component, message);
+    }
```
Please use constant strings and do not use magic strings for stuff like this.

metalgearsloth commented on Content.Shared/Access/Components/IdCardComponent.cs:
```diff
@@ -41,9 +41,17 @@ public sealed partial class IdCardComponent : Component
     [DataField, ViewVariables(VVAccess.ReadWrite)]
     public bool BypassLogging;
 
+
+    ///<summary>
+    /// Id used for the PDA messages system.
+    ///</summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public int MessagesId;
+
```
Please just use a struct for this.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+    {
+        return component.NameDict;
+    }
+
+    /// <summary>
+    /// Returns list of messages between the two users
+    /// </summary>
+    public List<MessagesMessageData> GetMessages(MessagesServerComponent component, int id1, int id2)
+    {
+        return new List<MessagesMessageData>(component.Messages.Where(message => message.SenderId == id1 && message.ReceiverId == id2 || message.SenderId == id2 && message.ReceiverId == id1));
+    }
+
```
This is going to be extremely slow to the point you're better off just using 2 dictionaries of senders and receivers on top of the message list.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+
+        _deviceNetworkSystem.QueuePacket(uid, null, packet);
+    }
+
+    /// <summary>
+    /// Returns the name of a given user
+    /// </summary>
+    public string GetNameFromDict(MessagesServerComponent component, int key)
+    {
+        if (component.NameDict.TryGetValue(key, out var value))
+            return value;
+        return Loc.GetString("messages-pda-user-missing");
+    }
```
This should return a nullable string or be a tryget with a bool return because callers have no idea this failed.

metalgearsloth commented on Content.Server/Radio/EntitySystems/MessagesServerSystem.cs:
```diff
@@ -0,0 +1,129 @@
+
+        var packet = new NetworkPayload()
+        {
+            ["Message"] = message,
+            ["ServerComponent"] = component
+        };
+
```
As above this looks like a code smell that the component is being added to a payload.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+    /// </summary>
+    private void OnPacketReceived(EntityUid uid, MessagesCartridgeComponent component, DeviceNetworkPacketEvent args)
+    {
+        if (!TryComp(uid, out CartridgeComponent? cartComponent))
+            return;
+        if (args.Data.TryGetValue<MessagesServerComponent>("ServerComponent", out var server))
+            component.LastServer = server;
+        if (args.Data.TryGetValue<bool>("NameQuery", out var _))
+            SendName(uid, component, cartComponent, args.SenderAddress);
+        if (args.Data.TryGetValue<MessagesMessageData>("Message", out var message) && cartComponent.LoaderUid != null)
+        {
+            if (message.ReceiverId == GetUserUid(cartComponent))
+            {
+                var name = GetName(message.SenderId, component);
+
+                var subtitleString = Loc.GetString("messages-pda-notification-header");
+
+                _cartridgeLoaderSystem.SendNotification(
+                    cartComponent.LoaderUid.Value,
+                    $"{name} {subtitleString}",
+                    message.Content);
+            }
+        }
+
+        if (HasComp<CartridgeLoaderComponent>(cartComponent.LoaderUid))
+            UpdateUiState(uid, cartComponent.LoaderUid.Value, component);
+
```
As per elsewhere.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+            UpdateUiState(uid, cartComponent.LoaderUid.Value, component);
+
+    }
+
+    /// <summary>
+    /// Sends the user's name to the server cache.
+    /// </summary>
+    private void SendName(EntityUid uid, MessagesCartridgeComponent component, CartridgeComponent cartComponent, string? address)
+    {
+        string name = GetUserName(cartComponent);
+
+        var packet = new NetworkPayload()
+        {
+            ["UserId"] = GetUserUid(cartComponent),
+            ["NewName"] = GetUserName(cartComponent),
+        };
+        _deviceNetworkSystem.QueuePacket(uid, address, packet);
+    }
+
```
Why does this need a device network packet just to update the name in the server?

juliangiebel replied:
The message cartridge system shouldn't be accessing the accessing the messaging system directly at all but updating the names should probably be handled in a different way from how it is now. While still using device networking though.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+        _deviceNetworkSystem.QueuePacket(uid, address, packet);
+    }
+
+    /// <summary>
+    /// Retrieves the name of the given user from the last contacted server
+    /// </summary>
+    private string GetName(int key, MessagesCartridgeComponent component)
+    {
+        if (component.LastServer == null)
+            return Loc.GetString("messages-pda-connection-error");
+        return _messagesServerSystem.GetNameFromDict(component.LastServer, key);
+    }
```
As per the other getname function.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+            return null;
+        return idComponent.MessagesId;
+    }
+
+    /// <summary>
+    /// Returns the user's name and job title
+    /// </summary>
+    public string GetUserName(CartridgeComponent component)
+    {
+        var idComponent = GetIdCard(component);
+        string job;
+        string name;
+        if (idComponent == null)
+            return Loc.GetString("messages-pda-unknown-name");
+        if (idComponent.FullName != null)
+        {
+            name = idComponent.FullName;
+        }
+        else
+        {
+            name = Loc.GetString("messages-pda-unknown-name");
+        }
+        if (idComponent.JobTitle != null)
+        {
+            job = idComponent.JobTitle;
+        }
+        else
+        {
+            job = Loc.GetString("messages-pda-unknown-job");
+        }
+        return $"{name} ({job})";
+    }
+
```
Just use id system for this instead of re-implementing.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+        if (HasComp<CartridgeLoaderComponent>(Transform(uid).ParentUid))
+            UpdateUiState(uid, Transform(uid).ParentUid, component);
+    }
+
+    private void UpdateUiState(EntityUid uid, EntityUid loaderUid, MessagesCartridgeComponent? component)
+    {
+        if (!Resolve(uid, ref component))
+            return;
+        if (!TryComp(uid, out CartridgeComponent? cartComponent))
+            return;
+        MessagesUiState state;
+        MapId mapId = Transform(uid).MapID;
```
Having a required nullable component but then resolving it anyway makes no sense, either make it optional or non-nullable.

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+        return CompOrNull<IdCardComponent>(pdaComponent.ContainedId);
+
+    }
+
+    ///<summary>
+    ///Updates the ui state of a given cartridge
+    ///</summary>
+    public void ForceUpdate(EntityUid uid, MessagesCartridgeComponent component)
+    {
+        if (HasComp<CartridgeLoaderComponent>(Transform(uid).ParentUid))
+            UpdateUiState(uid, Transform(uid).ParentUid, component);
+    }
+
```
Resolving components is expensive so please don't just put it in twice instead of putting it in a local variable.

You also get the component in a completely different way in OnPacketReceived instead of unifying it which is going to cause problems down the line.

Also getting the cartridge loader from an entity should be on cartridgeloadersystem not here.

metalgearsloth commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiMessageEvent.cs:
```diff
@@ -0,0 +1,27 @@
+public sealed class MessagesUiMessageEvent : CartridgeMessageEvent
+{
+    public readonly MessagesUiAction Action;
+    public readonly int? UidInput;
```
Why is the receiver nullable? Also you should call this the receiver.

PolterTzi replied:
Good point on the confusing name, I've changed it to something hopefully better.
This may have been a bad decision, but I'm using this same event for pretty much all the actions in the program UI.
This value is used to store the ID of the chat for when the user switches between chats. It can be null, because that's what the back button sets it to and because it's not always used.
The string can be null, because it is not used when switching chats.

metalgearsloth commented on Content.Shared/CartridgeLoader/Cartridges/MessagesUiMessageEvent.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    public readonly MessagesUiAction Action;
+    public readonly int? UidInput;
+    public readonly string? StringInput;
```
Why is a message able to be null?

metalgearsloth commented on Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs:
```diff
@@ -0,0 +1,304 @@
+            );
+
+            state = new MessagesUiState(MessagesUiStateMode.UserList, userList, null);
+        }
+        else
+        {
+            List<MessagesMessageData> messageList = []; //Else, list messages from the current chat
+
+            foreach (var message in _messagesServerSystem.GetMessages(component.LastServer, component.ChatUid.Value, currentUserId.Value))
+            {
+                if (message.SenderId == component.ChatUid && message.ReceiverId == currentUserId || message.ReceiverId == component.ChatUid && message.SenderId == currentUserId)
+                {
+                    messageList.Add(message);
+                }
+            }
+
+            messageList.Sort
+            (
+                delegate (MessagesMessageData a, MessagesMessageData b)
+                {
+                    return TimeSpan.Compare(a.Time, b.Time);
+                }
+            );
+
+            List<(string, int?)> formattedMessageList = [];
+
+            foreach (var message in messageList)
+            {
+                var name = GetName(message.SenderId, component);
+                var stationTime = message.Time.Subtract(_gameTicker.RoundStartTimeSpan);
+                var content = $"{stationTime.ToString("\\[hh\\:mm\\:ss\\]")} {name}: {message.Content}";
+                formattedMessageList.Add((content, null));
+            }
+
+            state = new MessagesUiState(MessagesUiStateMode.Chat, formattedMessageList, GetName(component.ChatUid.Value, component));
+        }
+        _cartridgeLoaderSystem.UpdateCartridgeUiState(loaderUid, state);
+    }
+
+}
```
Why do we need to send entire UI states if the user changes who they're chatting too? Their message history isn't going to be changing so this is just going to waste data, especially if they decide to view it again.

You're better off just using component states + SessionSpecific instead of a BUI state, then dirtying the cartridge if it gets loaded or not. Then whenever the messages change you can just dirty the component instead of trying to add handlers around this like you did above.

PolterTzi commented:
Yeah, just gonna take the liberty of closing it for the time being.
